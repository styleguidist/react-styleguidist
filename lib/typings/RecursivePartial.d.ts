/**
 * In a custom config file you might only want to override some parameters
 * This is the usage of the recursive Partial
 * `interface Test{param:string, paramObject:{p1:number, p2:boolean}}`
 * becomes
 * `interface TestPartial{param?:string, paramObject?:{p1?:number, p2?:boolean}}`
 * where everything is optional
 */
export declare type RecursivePartial<T> = {
    [P in keyof T]?: T[P] extends (infer U)[] ? RecursivePartial<U>[] : T[P] extends (...args: unknown[]) => unknown ? T[P] : T[P] extends Record<string, unknown> ? RecursivePartial<T[P]> : T[P];
};
