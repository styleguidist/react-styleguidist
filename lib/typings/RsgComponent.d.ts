import { MethodDescriptor, PropDescriptor, TagProps } from 'react-docgen';
import { RequireItResult } from './RsgRequireItResult';
import { Example } from './RsgExample';
export declare type ExpandMode = 'expand' | 'collapse' | 'hide';
export interface BaseComponent {
    hasExamples?: boolean;
    name?: string;
    slug?: string;
    href?: string;
    filepath?: string;
    pathLine?: string;
    description?: string;
    exampleMode?: ExpandMode;
    usageMode?: ExpandMode;
}
export interface Component extends BaseComponent {
    visibleName?: string;
    props?: {
        displayName?: string;
        visibleName?: string;
        description?: string;
        methods?: MethodDescriptor[];
        props?: PropDescriptor[];
        tags?: TagProps;
        example?: Example[];
        examples?: Example[];
    };
    module?: number;
    metadata?: {
        tags?: string[];
    };
}
export interface LoaderComponent extends BaseComponent {
    module: RequireItResult;
    props: RequireItResult;
    metadata: RequireItResult | Record<string, unknown>;
}
