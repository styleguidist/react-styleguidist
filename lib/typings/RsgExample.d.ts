export interface MarkdownExample {
    type: 'markdown';
    content: string;
    settings?: Record<string, any>;
}
export interface CodeExample {
    type: 'code';
    content: string;
    lang?: string | null;
    settings?: Record<string, any>;
}
export interface RuntimeCodeExample extends CodeExample {
    evalInContext(a: string): () => any;
}
export declare type Example = RuntimeCodeExample | MarkdownExample;
