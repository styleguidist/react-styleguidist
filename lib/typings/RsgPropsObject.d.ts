import { DocumentationObject, MethodDescriptor, PropDescriptor } from 'react-docgen';
import { RequireItResult } from './RsgRequireItResult';
export interface MethodWithDocblock extends MethodDescriptor {
    docblock: string;
}
export interface TempPropsObject extends DocumentationObject {
    displayName: string;
    visibleName?: string;
    methods?: MethodWithDocblock[];
    doclets: Record<string, any>;
    example?: RequireItResult | null;
}
export interface PropsObject extends Omit<TempPropsObject, 'props'> {
    props?: Record<string, PropDescriptor> | PropDescriptor[];
    examples?: RequireItResult | null;
}
