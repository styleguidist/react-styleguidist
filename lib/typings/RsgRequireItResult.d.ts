import { ASTNode } from 'ast-types';
export interface RequireItResult {
    require: string;
    toAST(): ASTNode;
}
