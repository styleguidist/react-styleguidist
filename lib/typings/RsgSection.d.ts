/// <reference types="react" />
import { RequireItResult } from './RsgRequireItResult';
import { MarkdownExample, Example } from './RsgExample';
import { LoaderComponent, ExpandMode, Component } from './RsgComponent';
export interface BaseSection {
    name?: string;
    slug?: string;
    ignore?: string | string[];
    description?: string;
    exampleMode?: ExpandMode;
    usageMode?: ExpandMode;
    href?: string;
    sectionDepth?: number;
    external?: boolean;
    expand?: boolean;
}
export interface ProcessedSection extends BaseSection {
    visibleName?: string;
    filepath?: string;
    externalLink?: boolean;
    href?: string;
}
/**
 * Section used on the client in javascript
 * It is the output of the function `client/utils/processSection`
 */
export interface Section extends ProcessedSection {
    content?: Example[] | string;
    components?: Component[];
    sections?: Section[];
}
/**
 * Item of the Table Of Contents used in
 * ComponentsList
 * TableOfContent
 * filterSectionByName
 */
export interface TOCItem extends ProcessedSection {
    heading?: boolean;
    shouldOpenInNewTab?: boolean;
    selected?: boolean;
    initialOpen?: boolean;
    forcedOpen?: boolean;
    content?: React.ReactNode;
    components?: TOCItem[];
    sections?: TOCItem[];
}
/**
 * Used in the config file and at the early stages of processing
 * in `schema/config.ts` this is the type that is used
 */
export interface ConfigSection extends BaseSection {
    components?: string | string[] | (() => string[]);
    sections?: ConfigSection[];
    content?: string;
}
/**
 * Type returned when sections are transformed to their webpack
 * loadable equivalents
 */
export interface LoaderSection extends BaseSection {
    slug?: string;
    content?: RequireItResult | MarkdownExample;
    components: LoaderComponent[];
    sections: LoaderSection[];
}
