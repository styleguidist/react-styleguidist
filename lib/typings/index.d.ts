import './dependencies/acorn-jsx';
import './dependencies/findup';
import './dependencies/listify';
import './dependencies/react-docgen';
import './dependencies/webpack-merge';
import './dependencies/common-dir';
import './dependencies/github-slugger';
import './dependencies/strip-shebang';
import './dependencies/deabsdeep';
import './dependencies/glogg';
import './dependencies/mini-html-webpack-template';
import './dependencies/stripHtmlComments';
import './dependencies/deepfreeze';
import './dependencies/is-directory';
import './dependencies/q-i';
import './dependencies/to-ast';
export * from './RsgComponent';
export * from './RsgExample';
export * from './RsgPropsObject';
export * from './RsgRequireItResult';
export * from './RsgSection';
export * from './RsgStyleguidistConfig';
export * from './RsgTheme';
