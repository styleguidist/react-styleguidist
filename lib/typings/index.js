"use strict";

exports.__esModule = true;

require("./dependencies/acorn-jsx");

require("./dependencies/findup");

require("./dependencies/listify");

require("./dependencies/react-docgen");

require("./dependencies/webpack-merge");

require("./dependencies/common-dir");

require("./dependencies/github-slugger");

require("./dependencies/strip-shebang");

require("./dependencies/deabsdeep");

require("./dependencies/glogg");

require("./dependencies/mini-html-webpack-template");

require("./dependencies/stripHtmlComments");

require("./dependencies/deepfreeze");

require("./dependencies/is-directory");

require("./dependencies/q-i");

require("./dependencies/to-ast");

var _RsgComponent = require("./RsgComponent");

Object.keys(_RsgComponent).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  exports[key] = _RsgComponent[key];
});

var _RsgExample = require("./RsgExample");

Object.keys(_RsgExample).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  exports[key] = _RsgExample[key];
});

var _RsgPropsObject = require("./RsgPropsObject");

Object.keys(_RsgPropsObject).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  exports[key] = _RsgPropsObject[key];
});

var _RsgRequireItResult = require("./RsgRequireItResult");

Object.keys(_RsgRequireItResult).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  exports[key] = _RsgRequireItResult[key];
});

var _RsgSection = require("./RsgSection");

Object.keys(_RsgSection).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  exports[key] = _RsgSection[key];
});

var _RsgStyleguidistConfig = require("./RsgStyleguidistConfig");

Object.keys(_RsgStyleguidistConfig).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  exports[key] = _RsgStyleguidistConfig[key];
});

var _RsgTheme = require("./RsgTheme");

Object.keys(_RsgTheme).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  exports[key] = _RsgTheme[key];
});