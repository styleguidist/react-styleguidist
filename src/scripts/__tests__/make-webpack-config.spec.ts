import webpack, { Configuration } from 'webpack';
import { Tapable } from 'tapable';
import CopyWebpackPlugin from 'copy-webpack-plugin';
import makeWebpackConfig from '../make-webpack-config';
import * as Rsg from '../../typings';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const { validate } = require('webpack');

jest.mock('copy-webpack-plugin');

const styleguideConfig = ({
	styleguideDir: __dirname,
	require: [],
	title: 'Style Guide',
} as unknown) as Rsg.SanitizedStyleguidistConfig;

const getClasses = (plugins: Tapable.Plugin[] = [], name: string): Tapable.Plugin[] =>
	plugins.filter((x) => x.constructor.name === name);
const getClassNames = (plugins: Tapable.Plugin[] = []): string[] =>
	plugins.map((x) => x.constructor.name);

const process$env$nodeEnv = process.env.NODE_ENV;

beforeEach(() => {
	(CopyWebpackPlugin as jest.Mock).mockClear();
});

afterEach(() => {
	process.env.NODE_ENV = process$env$nodeEnv;
});

it('should return a development config', () => {
	const env = 'development';
	const config = makeWebpackConfig(styleguideConfig, env);

	const errors = validate(config);
	expect(errors).toHaveLength(0);

	expect(config).toMatchObject({
		mode: env,
	});
	expect(config).not.toHaveProperty('optimization');
});

it('should return a production config', () => {
	const env = 'production';
	const config = makeWebpackConfig(styleguideConfig, env);
	const errors = validate(config);
	expect(errors).toHaveLength(0);

	const plugins = getClassNames(config.plugins);
	expect(plugins).toContain('CleanWebpackPlugin');
	expect(plugins).not.toContain('HotModuleReplacementPlugin');

	expect(config).toMatchObject({
		output: {
			filename: expect.stringContaining('[chunkhash'),
			chunkFilename: expect.stringContaining('[chunkhash'),
		},
	});

	expect(config).toMatchObject({
		mode: env,
	});
	expect(
		getClasses(config.optimization && config.optimization.minimizer, 'TerserPlugin')
	).toHaveLength(1);
});

it('should set aliases', () => {
	const result = makeWebpackConfig(styleguideConfig, 'development');
	expect(result.resolve && result.resolve.alias).toMatchSnapshot();
});

it('should set aliases from moduleAliases option', () => {
	const result = makeWebpackConfig(
		{
			...styleguideConfig,
			moduleAliases: {
				foo: 'bar',
			},
		},
		'development'
	);
	expect(result.resolve && result.resolve.alias).toMatchSnapshot();
});

it('should set aliases from styleguideComponents option', () => {
	const result = makeWebpackConfig(
		{
			...styleguideConfig,
			styleguideComponents: {
				foo: 'bar',
			},
		},
		'development'
	);
	expect(result.resolve && result.resolve.alias).toMatchSnapshot();
});

it('should prepend requires as webpack entries', () => {
	const result = makeWebpackConfig(
		{ ...styleguideConfig, require: ['a/b.js', 'c/d.css'] },
		'development'
	);
	expect(result.entry).toMatchSnapshot();
});

it('should enable verbose mode in CleanWebpackPlugin', () => {
	const result = makeWebpackConfig({ ...styleguideConfig, verbose: true }, 'production');
	expect((getClasses(result.plugins, 'CleanWebpackPlugin')[0] as any).verbose).toBe(true);
});

it('should set from with assetsDir in CopyWebpackPlugin', () => {
	makeWebpackConfig({ ...styleguideConfig, assetsDir: '/assets/' }, 'production');
	expect(CopyWebpackPlugin).toHaveBeenCalledWith({
		patterns: [{ from: '/assets/' }],
	});
});

it('should set array of from with assetsDir array in CopyWebpackPlugin', () => {
	makeWebpackConfig({ ...styleguideConfig, assetsDir: ['/assets1/', '/assets2/'] }, 'production');
	expect(CopyWebpackPlugin).toHaveBeenCalledWith({
		patterns: [{ from: '/assets1/' }, { from: '/assets2/' }],
	});
});

it('should merge user webpack config', () => {
	const result = makeWebpackConfig(
		{ ...styleguideConfig, webpackConfig: { resolve: { alias: { foo: 'bar' } } } },
		'development'
	);
	expect(result.resolve && result.resolve.alias).toMatchSnapshot();
});

it('should not owerwrite user DefinePlugin', () => {
	const result = makeWebpackConfig(
		{
			...styleguideConfig,
			webpackConfig: {
				plugins: [
					new webpack.DefinePlugin({
						'process.env.PIZZA': JSON.stringify('salami'),
					}),
				],
			},
		},
		'development'
	);

	// Doesn’t really test that values won’t be overwritten, just that
	// DefinePlugin is applied twice. To write a real test we’d have to run
	// webpack
	expect(getClasses(result.plugins, 'DefinePlugin')).toMatchSnapshot();
});

it('should update webpack config', () => {
	const extensions = ['.web.js', '.js'];
	const result = makeWebpackConfig(
		{
			...styleguideConfig,
			dangerouslyUpdateWebpackConfig: (c: Configuration) => {
				if (c.resolve) {
					c.resolve.extensions = extensions;
				}
				return c;
			},
		},
		'development'
	);
	expect(result.resolve && result.resolve.extensions).toEqual(extensions);
});

it('should pass template context to HTML plugin', () => {
	const template = {
		pizza: 'salami',
	};
	const result = makeWebpackConfig(
		{
			...styleguideConfig,
			template,
		},
		'development'
	);
	expect(getClasses(result.plugins, 'MiniHtmlWebpackPlugin')[0]).toMatchObject({
		options: {
			context: template,
			template: expect.any(Function),
		},
	});
});

it('should pass template function to HTML plugin', () => {
	const template = () => '<html />';
	const result = makeWebpackConfig(
		{
			...styleguideConfig,
			template,
		},
		'development'
	);
	expect(getClasses(result.plugins, 'MiniHtmlWebpackPlugin')[0]).toMatchObject({
		options: {
			context: expect.any(Object),
			template,
		},
	});
});

it('should update NODE_ENV', () => {
	process.env.NODE_ENV = '';
	makeWebpackConfig(styleguideConfig, 'production');
	expect(process.env.NODE_ENV).toBe('production');
});

it('should not overwrite NODE_ENV', () => {
	makeWebpackConfig(styleguideConfig, 'production');
	expect(process.env.NODE_ENV).toBe(process$env$nodeEnv);
});

it('should pass specified mountPointId to HTML plugin', () => {
	const result = makeWebpackConfig(
		{
			...styleguideConfig,
			mountPointId: 'foo-bar',
		},
		'development'
	);
	expect(
		(getClasses(result.plugins, 'MiniHtmlWebpackPlugin')[0] as any).options.context.container
	).toEqual('foo-bar');
});
