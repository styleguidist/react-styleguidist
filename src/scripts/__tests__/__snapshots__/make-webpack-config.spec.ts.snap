// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should merge user webpack config 1`] = `
Object {
  "foo": "bar",
  "rsg-components": "~/src/client/rsg-components",
}
`;

exports[`should not owerwrite user DefinePlugin 1`] = `
Array [
  DefinePlugin {
    "definitions": Object {
      "process.env.NODE_ENV": "\\"test\\"",
      "process.env.STYLEGUIDIST_ENV": "\\"development\\"",
    },
  },
  DefinePlugin {
    "definitions": Object {
      "process.env.PIZZA": "\\"salami\\"",
    },
  },
]
`;

exports[`should prepend requires as webpack entries 1`] = `
Array [
  "a/b.js",
  "c/d.css",
  "~/src/client/index",
]
`;

exports[`should set aliases 1`] = `
Object {
  "rsg-components": "~/src/client/rsg-components",
}
`;

exports[`should set aliases from moduleAliases option 1`] = `
Object {
  "foo": "bar",
  "rsg-components": "~/src/client/rsg-components",
}
`;

exports[`should set aliases from styleguideComponents option 1`] = `
Object {
  "rsg-components": "~/src/client/rsg-components",
  "rsg-components/foo": "bar",
}
`;
